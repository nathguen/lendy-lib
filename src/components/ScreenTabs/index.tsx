import React, { useState } from 'react';
import { StyleSheet } from 'react-native';
import { TouchableOpacity, View } from 'react-native';

import Colors from '../../common/colors';
import Text from '../Text';

interface ScreenTabsProps {
  tabs: string[];
  defaultTab?: string;
  onChangeTab?: (tab: string) => void;
}

const ScreenTabs = ({ tabs, defaultTab, onChangeTab }: ScreenTabsProps) => {
  const [selectedTab, setSelectedTab] = useState(defaultTab || tabs[0]);

  const handleTabChange = (tab: string) => {
    setSelectedTab(tab);

    if (onChangeTab && typeof onChangeTab === 'function') {
      onChangeTab(tab);
    }
  };

  return (
    <View style={styles.tabsWrapper}>
      {tabs.map((tab) => (
        <TouchableOpacity
          key={tab}
          onPress={() => handleTabChange(tab)}
          style={styles.tabPressable}
        >
          <Text
            style={{
              ...styles.tabText,
              ...(tab === selectedTab && styles.tabTextActive),
            }}
          >
            {tab}
          </Text>

          <View
            style={{
              ...styles.tabUnderline,
              ...(tab === selectedTab && styles.tabUnderlineActive),
            }}
          />
        </TouchableOpacity>
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  tabsWrapper: {
    flexDirection: 'row',
    width: '100%',
    justifyContent: 'center',
  },
  tabPressable: {
    paddingVertical: 16,
    paddingHorizontal: 32,
    alignItems: 'center',
  },
  tabText: {
    marginBottom: 4,
    color: Colors.Gray,
  },
  tabTextActive: {
    color: Colors.Dark,
  },
  tabUnderline: {
    width: 42,
    borderColor: Colors.Yellow,
    borderWidth: 0,
    borderRadius: 2,
  },
  tabUnderlineActive: {
    borderWidth: 2,
  },
});

export default ScreenTabs;
