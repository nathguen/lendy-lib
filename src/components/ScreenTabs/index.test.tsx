import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import ScreenTabs, { getScreenTabsTestId } from './';

describe('ScreenTabs', () => {
  describe('2 tabs', () => {
    it('renders the tabs', () => {
      const tabs: string[] = ['Tab 1', 'Tab 2'];

      const { toJSON, queryByTestId, getByTestId } = render(
        <ScreenTabs tabs={tabs} />
      );

      expect(toJSON()).toMatchSnapshot();
      expect(getByTestId(getScreenTabsTestId('Tab 1', true))).toBeTruthy();
      expect(queryByTestId(getScreenTabsTestId('Tab 2', false))).toBeTruthy();
    });

    it('renders the tabs and defaults to the second one', () => {
      const tabs: string[] = ['Tab 1', 'Tab 2'];

      const { toJSON: initialToJSON } = render(<ScreenTabs tabs={tabs} />);
      const { toJSON: secondToJSON } = render(
        <ScreenTabs tabs={tabs} defaultTab={'Tab 2'} />
      );

      expect(initialToJSON()).not.toEqual(secondToJSON());
    });

    it('renders the tabs and switches to second one when pressed', () => {
      const tabs: string[] = ['Tab 1', 'Tab 2'];
      const mockedOnTabPress = jest.fn();

      const { getByText, getByTestId, queryByTestId } = render(
        <ScreenTabs tabs={tabs} onChangeTab={mockedOnTabPress} />
      );

      fireEvent.press(getByText('Tab 2'));
      expect(mockedOnTabPress).toHaveBeenCalledWith('Tab 2');
      expect(getByTestId(getScreenTabsTestId('Tab 2', true))).toBeTruthy();
      expect(queryByTestId(getScreenTabsTestId('Tab 1', false))).toBeTruthy();
    });
  });

  describe('3 tabs', () => {
    it('renders the tabs and defaults to the first one without "defaultTab" prop', () => {
      const tabs: string[] = ['Tab 1', 'Tab 2', 'Tab 3'];

      const { toJSON, queryByTestId } = render(<ScreenTabs tabs={tabs} />);

      expect(toJSON()).toMatchSnapshot();
      expect(queryByTestId(getScreenTabsTestId('Tab 1', true))).toBeTruthy();
    });

    it('renders the tabs and defaults to the third one', () => {
      const tabs: string[] = ['Tab 1', 'Tab 2', 'Tab 3'];

      const { toJSON, queryByTestId } = render(
        <ScreenTabs tabs={tabs} defaultTab="Tab 3" />
      );

      expect(toJSON()).toMatchSnapshot();
      expect(queryByTestId(getScreenTabsTestId('Tab 3', true))).toBeTruthy();
    });

    it('renders the tabs and switches to third one when pressed', () => {
      const tabs: string[] = ['Tab 1', 'Tab 2', 'Tab 3'];
      const mockedOnTabPress = jest.fn();

      const { getByText, getByTestId, queryByTestId } = render(
        <ScreenTabs tabs={tabs} onChangeTab={mockedOnTabPress} />
      );

      fireEvent.press(getByText('Tab 3'));
      expect(mockedOnTabPress).toHaveBeenCalledWith('Tab 3');
      expect(getByTestId(getScreenTabsTestId('Tab 3', true))).toBeTruthy();
      expect(queryByTestId(getScreenTabsTestId('Tab 1', false))).toBeTruthy();
      expect(queryByTestId(getScreenTabsTestId('Tab 2', false))).toBeTruthy();
    });
  });
});
