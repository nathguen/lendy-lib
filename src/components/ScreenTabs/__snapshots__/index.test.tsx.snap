// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenTabs 2 tabs renders the tabs 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabSelected::Tab 1"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 1
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 2,
          "width": 42,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabUnselected::Tab 2"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 2
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 0,
          "width": 42,
        }
      }
    />
  </View>
</View>
`;

exports[`ScreenTabs 3 tabs renders the tabs and defaults to the first one without "defaultTab" prop 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabSelected::Tab 1"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 1
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 2,
          "width": 42,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabUnselected::Tab 2"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 2
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 0,
          "width": 42,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabUnselected::Tab 3"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 3
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 0,
          "width": 42,
        }
      }
    />
  </View>
</View>
`;

exports[`ScreenTabs 3 tabs renders the tabs and defaults to the third one 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabUnselected::Tab 1"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 1
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 0,
          "width": 42,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabUnselected::Tab 2"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 2
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 0,
          "width": 42,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "opacity": 1,
        "paddingHorizontal": 32,
        "paddingVertical": 16,
      }
    }
    testID="ScreenTabSelected::Tab 3"
  >
    <Text
      style={
        Object {
          "color": "#191919",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 4,
        }
      }
    >
      Tab 3
    </Text>
    <View
      style={
        Object {
          "borderColor": "#ECC850",
          "borderRadius": 2,
          "borderWidth": 2,
          "width": 42,
        }
      }
    />
  </View>
</View>
`;
